File: memorezar-android/app/src/main/java/com/memorezar/app/data/services/UnsplashService.kt
package com.memorezar.app.data.services

import android.util.Log
import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.request.get
import io.ktor.client.request.header
import io.ktor.http.isSuccess
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import javax.inject.Inject
import javax.inject.Singleton

private const val TAG = "UnsplashService"
private const val BASE_URL = "https://api.unsplash.com"
private const val ACCESS_KEY = "7YzF0aG-BRdz4vAQBXSIEfOvYbaC4hIC5Eu7-lmfxwc"

// Response models

data class UnsplashImageInfo(
    val regularURL: String,
    val smallURL: String,
    val photographerName: String,
    val photographerURL: String,
    val downloadURL: String
)

@Serializable
private data class UnsplashUrls(
    val regular: String = "",
    val small: String = ""
)

@Serializable
private data class UnsplashUserLinks(
    val html: String = ""
)

@Serializable
private data class UnsplashUser(
    val name: String = "",
    val links: UnsplashUserLinks = UnsplashUserLinks()
)

@Serializable
private data class UnsplashPhotoLinks(
    @SerialName("download_location") val downloadLocation: String = ""
)

@Serializable
private data class UnsplashPhoto(
    val urls: UnsplashUrls = UnsplashUrls(),
    val user: UnsplashUser = UnsplashUser(),
    val links: UnsplashPhotoLinks = UnsplashPhotoLinks()
)

@Serializable
private data class UnsplashSearchResponse(
    val results: List<UnsplashPhoto> = emptyList()
)

// Errors

sealed class UnsplashError(override val message: String) : Exception(message) {
    object RateLimited : UnsplashError("Unsplash rate limit exceeded")
    data class HttpError(val statusCode: Int) : UnsplashError("Unsplash HTTP $statusCode")
    object InvalidResponse : UnsplashError("Invalid Unsplash response")
}

@Singleton
class UnsplashService @Inject constructor(
    private val httpClient: HttpClient
) {
    private val json = Json { ignoreUnknownKeys = true }

    suspend fun fetchRandomPhoto(query: String? = null): UnsplashImageInfo {
        val url = buildString {
            append("$BASE_URL/photos/random?orientation=portrait")
            if (!query.isNullOrBlank()) append("&query=$query")
        }

        val response = httpClient.get(url) {
            header("Authorization", "Client-ID $ACCESS_KEY")
        }

        if (response.status.value == 403) throw UnsplashError.RateLimited
        if (!response.status.isSuccess()) throw UnsplashError.HttpError(response.status.value)

        val body: String = response.body()
        val photo = json.decodeFromString<UnsplashPhoto>(body)
        return photo.toImageInfo()
    }

    suspend fun searchPhotos(query: String, page: Int = 1): List<UnsplashImageInfo> {
        val url = "$BASE_URL/search/photos?query=$query&page=$page&per_page=20&orientation=portrait"

        val response = httpClient.get(url) {
            header("Authorization", "Client-ID $ACCESS_KEY")
        }

        if (response.status.value == 403) throw UnsplashError.RateLimited
        if (!response.status.isSuccess()) throw UnsplashError.HttpError(response.status.value)

        val body: String = response.body()
        val searchResponse = json.decodeFromString<UnsplashSearchResponse>(body)
        return searchResponse.results.map { it.toImageInfo() }
    }

    /** Trigger download event per Unsplash ToS (fire-and-forget). */
    suspend fun triggerDownload(url: String) {
        try {
            httpClient.get(url) {
                header("Authorization", "Client-ID $ACCESS_KEY")
            }
        } catch (_: Exception) { }
    }

    private fun UnsplashPhoto.toImageInfo() = UnsplashImageInfo(
        regularURL = urls.regular,
        smallURL = urls.small,
        photographerName = user.name,
        photographerURL = user.links.html,
        downloadURL = links.downloadLocation
    )
}

File: ios/Memorezar/Data/Services/UnsplashService.swift
import Foundation

/// Async service for fetching photos from the Unsplash API
final class UnsplashService {
    static let shared = UnsplashService()
    private init() {}

    // Free tier: 50 requests/hour
    private let accessKey = "7YzF0aG-BRdz4vAQBXSIEfOvYbaC4hIC5Eu7-lmfxwc"
    private let baseURL = "https://api.unsplash.com"
    private let session = URLSession.shared

    // MARK: - Public API

    /// Fetch a single random photo, optionally filtered by query
    func fetchRandomPhoto(query: String? = nil) async throws -> UnsplashImageInfo {
        var components = URLComponents(string: "\(baseURL)/photos/random")!
        var queryItems = [URLQueryItem(name: "orientation", value: "portrait")]
        if let query = query {
            queryItems.append(URLQueryItem(name: "query", value: query))
        }
        components.queryItems = queryItems

        let data = try await performRequest(url: components.url!)
        let photo = try JSONDecoder().decode(UnsplashPhoto.self, from: data)
        return photo.toImageInfo()
    }

    /// Search photos by query with pagination
    func searchPhotos(query: String, page: Int = 1) async throws -> [UnsplashImageInfo] {
        var components = URLComponents(string: "\(baseURL)/search/photos")!
        components.queryItems = [
            URLQueryItem(name: "query", value: query),
            URLQueryItem(name: "page", value: "\(page)"),
            URLQueryItem(name: "per_page", value: "20"),
            URLQueryItem(name: "orientation", value: "portrait"),
        ]

        let data = try await performRequest(url: components.url!)
        let response = try JSONDecoder().decode(UnsplashSearchResponse.self, from: data)
        return response.results.map { $0.toImageInfo() }
    }

    /// Trigger a download event per Unsplash API TOS
    func triggerDownload(url: String) async {
        guard let downloadURL = URL(string: url) else { return }
        var request = URLRequest(url: downloadURL)
        request.setValue("Client-ID \(accessKey)", forHTTPHeaderField: "Authorization")
        _ = try? await session.data(for: request)
    }

    // MARK: - Private

    private func performRequest(url: URL) async throws -> Data {
        var request = URLRequest(url: url)
        request.setValue("Client-ID \(accessKey)", forHTTPHeaderField: "Authorization")
        request.timeoutInterval = 10

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw UnsplashError.invalidResponse
        }

        guard (200...299).contains(httpResponse.statusCode) else {
            if httpResponse.statusCode == 403 {
                throw UnsplashError.rateLimited
            }
            throw UnsplashError.httpError(httpResponse.statusCode)
        }

        return data
    }
}

// MARK: - Error Types

enum UnsplashError: Error, LocalizedError {
    case invalidResponse
    case rateLimited
    case httpError(Int)

    var errorDescription: String? {
        switch self {
        case .invalidResponse: return "Invalid response from Unsplash"
        case .rateLimited: return "Unsplash rate limit reached. Try again later."
        case .httpError(let code): return "Unsplash API error (HTTP \(code))"
        }
    }
}

// MARK: - API Response Models

private struct UnsplashPhoto: Decodable {
    let urls: UnsplashURLs
    let user: UnsplashUser
    let links: UnsplashLinks

    func toImageInfo() -> UnsplashImageInfo {
        UnsplashImageInfo(
            regularURL: urls.regular,
            smallURL: urls.small,
            photographerName: user.name,
            photographerURL: user.links.html,
            downloadURL: links.download_location
        )
    }
}

private struct UnsplashURLs: Decodable {
    let regular: String
    let small: String
}

private struct UnsplashUser: Decodable {
    let name: String
    let links: UnsplashUserLinks
}

private struct UnsplashUserLinks: Decodable {
    let html: String
}

private struct UnsplashLinks: Decodable {
    let download_location: String
}

private struct UnsplashSearchResponse: Decodable {
    let results: [UnsplashPhoto]
}
